src/libAtomVM/nifs.c

```diff
@@ -3165,6 +3172,133 @@ static term nif_binary_split(Context *ctx, int argc, term argv[])
+        term pattern_term = term_get_list_head(t);
+        if (UNLIKELY(!term_is_binary(pattern_term))) {
+            return false;
+        }
+        if (UNLIKELY(term_binary_size(pattern_term) == 0)) {
+            return false;
+        }
+        t = term_get_list_tail(t);
+    }
+    bool proper = term_is_nil(t);
+    if (UNLIKELY(!proper)) {
+        return false;
+    }
+    return true;
+}
+
+static BinaryPosLen find_pattern_in_binary(term binary_term, BinaryPosLen scope_slice, term pattern_term)
+{
+    const char *binary = term_binary_data(binary_term) + scope_slice.pos;
+    size_t size = scope_slice.len;
+    const char *pattern = term_binary_data(pattern_term);
+    size_t pattern_size = term_binary_size(pattern_term);
+
+    BinaryPosLen pattern_slice = term_invalid_binary_pos_len();
+    const char *sub_binary = memmem(binary, size, pattern, pattern_size);
+    if (sub_binary != NULL) {
+        pattern_slice.len = pattern_size;
+        pattern_slice.pos = (sub_binary - binary) + scope_slice.pos;
+    }
+    return pattern_slice;
+}
+
+static BinaryPosLen select_earlier_slice(BinaryPosLen old_slice, BinaryPosLen new_slice)
+{
+    if (term_is_invalid_binary_pos_len(new_slice)) {
+        return old_slice;
+    }
+    if (term_is_invalid_binary_pos_len(old_slice)) {
+        return new_slice;
+    }
+    if (new_slice.pos < old_slice.pos) {
+        return new_slice;
+    }
+    if (new_slice.pos == old_slice.pos && new_slice.len > old_slice.len) {
+        return new_slice;
+    }
+    return old_slice;
+}
+
+static term nif_binary_match(Context *ctx, int argc, term argv[])
+{
+    term scope_atom = globalcontext_make_atom(ctx->global, ATOM_STR("\x5", "scope"));
+    term nomatch_atom = globalcontext_make_atom(ctx->global, ATOM_STR("\x7", "nomatch"));
+    term binary_term = argv[0];
+    term pattern_or_patterns_term = argv[1];
+    term options_term = argc == 3 ? argv[2] : term_nil();
+
+    VALIDATE_VALUE(binary_term, term_is_binary);
+    VALIDATE_VALUE(options_term, term_is_list);
+    VALIDATE_VALUE(pattern_or_patterns_term, is_valid_pattern);
+
+    BinaryPosLen scope_slice;
+    if (UNLIKELY(!get_binary_scope_slice(binary_term, options_term, scope_atom, &scope_slice))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    BinaryPosLen match_slice = term_invalid_binary_pos_len();
+    if (term_is_binary(pattern_or_patterns_term)) {
+        term pattern_term = pattern_or_patterns_term;
+        match_slice = find_pattern_in_binary(binary_term, scope_slice, pattern_term);
+    } else {
+        term patterns = pattern_or_patterns_term;
+        while (!term_is_nil(patterns)) {
+            term pattern_term = term_get_list_head(patterns);
+            BinaryPosLen new_match_slice = find_pattern_in_binary(binary_term, scope_slice, pattern_term);
+            match_slice = select_earlier_slice(match_slice, new_match_slice);
+            patterns = term_get_list_tail(patterns);
+        }
```

Comment:
Let's raise here at the end if patterns is not a list. (that means that it is improper).

Reply:
This is not necessary, we check that first in `VALIDATE_VALUE(pattern_or_patterns_term, is_valid_pattern);`.